File: CustomApplications/VB.Net/GraphicsHowTo/ScreenOverlays/OverlaysPanelCodeSnippet.vb
#Region "UsingDirectives"

Imports System.IO
Imports System.Drawing
Imports GraphicsHowTo.Imaging
Imports AGI.STKGraphics
Imports AGI.STKObjects

#End Region

Namespace ScreenOverlays
	Public Class OverlaysPanelCodeSnippet
		Inherits CodeSnippet
		Public Sub New()
            MyBase.New("ScreenOverlays\OverlaysPanelCodeSnippet.vb")
        End Sub

        Public Overrides Sub Execute(ByVal scene As IAgStkGraphicsScene, ByVal root As AgStkObjectRoot)
            Dim imageFile As String = New AGI.DataPath(AGI.DataPathRoot.Relative, "Markers/originalLogo.png").FullPath
            Dim rasterFile As String = New AGI.DataPath(AGI.DataPathRoot.Relative, "Textures/SpinSat_256.gif").FullPath
            ExecuteSnippet(scene, root, imageFile, rasterFile)
        End Sub

		' Name        
		' Description 
		' Category    
		' References  
		' Namespaces  
		' EID         
        <AGI.CodeSnippets.CodeSnippet( _
            "AddAPanelOverlay", _
            "Add overlays to a panel overlay", _
            "Graphics | ScreenOverlays", _
            "System", _
            "System", _
            "AgSTKGraphicsLib~IAgStkGraphicsScreenOverlay" _
            )> _
        Public Sub ExecuteSnippet(<AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")> ByVal scene As IAgStkGraphicsScene, <AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")> ByVal root As AgStkObjectRoot, <AGI.CodeSnippets.CodeSnippet.Parameter("imageFile", "The image file to use for the child overlay")> ByVal imageFile As String, <AGI.CodeSnippets.CodeSnippet.Parameter("rasterFile", "The raster file for the second child overlay")> ByVal rasterFile As String)
            Try
                '#Region "CodeSnippet"
                Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager
                Dim overlayManager As IAgStkGraphicsScreenOverlayCollectionBase = DirectCast(manager.ScreenOverlays.Overlays, IAgStkGraphicsScreenOverlayCollectionBase)

                Dim overlay As IAgStkGraphicsTextureScreenOverlay = manager.Initializers.TextureScreenOverlay.InitializeWithXYWidthHeight(attachID("$xLocation$The x location of the screen overlay$", 0), attachID("$yLocation$The y location of the screen overlay$", 0), attachID("$overlayWidth$The width of the screen overlay$", 188), attachID("$overlayHeight$The height of the screen overlay$", 200))
                DirectCast(overlay, IAgStkGraphicsOverlay).Origin = attachID("$origin$The origin of the screen overlay$", AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginTopLeft)
                DirectCast(overlay, IAgStkGraphicsOverlay).BorderTranslucency = attachID("$borderTranslucency$The translucency of the border$", 0.3F)
                DirectCast(overlay, IAgStkGraphicsOverlay).BorderSize = attachID("$borderWidth$The width of the border$", 1)
                DirectCast(overlay, IAgStkGraphicsOverlay).BorderColor = attachID("$borderColor$The System.Drawing.Color of the border$", Color.LightBlue)
                DirectCast(overlay, IAgStkGraphicsOverlay).Color = attachID("$color$The System.Drawing.Color of the screen overlay$", Color.LightSkyBlue)
                DirectCast(overlay, IAgStkGraphicsOverlay).Translucency = attachID("$translucency$The translucency of the screen overlay$", 0.7F)

                Dim childOverlay As IAgStkGraphicsTextureScreenOverlay = manager.Initializers.TextureScreenOverlay.InitializeWithXYWidthHeight(attachID("$childXLocation$The x location of the child overlay$", 0), attachID("$childYLocation$The y location of the child overlay$", 0), DirectCast(overlay, IAgStkGraphicsOverlay).Width, DirectCast(overlay, IAgStkGraphicsOverlay).Height)
                DirectCast(childOverlay, IAgStkGraphicsOverlay).Origin = attachID("$childOrigin$The origin of the child overlay$", AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginCenter)

                childOverlay.Texture = manager.Textures.LoadFromStringUri(imageFile)

                '
                ' Create the RasterStream from the plugin the same way as in ImageDynamicCodeSnippet.cs
                '
                Dim activator As IAgStkGraphicsProjectionRasterStreamPluginActivator = manager.Initializers.ProjectionRasterStreamPluginActivator.Initialize()
                Dim proxy As IAgStkGraphicsProjectionRasterStreamPluginProxy = activator.CreateFromDisplayName("ProjectionRasterStreamPlugin.VBNET")
                Dim plugin As Type = proxy.RealPluginObject.[GetType]()
                plugin.GetProperty("RasterPath").SetValue(proxy.RealPluginObject, rasterFile, Nothing)

                Dim rasterStream As IAgStkGraphicsRasterStream = proxy.RasterStream
                rasterStream.UpdateDelta = 0.01667
                Dim texture2D As IAgStkGraphicsRendererTexture2D = manager.Textures.FromRaster(DirectCast(rasterStream, IAgStkGraphicsRaster))

                Dim secondChildOverlay As IAgStkGraphicsTextureScreenOverlay = manager.Initializers.TextureScreenOverlay.InitializeWithXYWidthHeight(0, 0, 128, 128)
                DirectCast(secondChildOverlay, IAgStkGraphicsOverlay).Origin = AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginTopRight
                DirectCast(secondChildOverlay, IAgStkGraphicsOverlay).TranslationX = -36
                DirectCast(secondChildOverlay, IAgStkGraphicsOverlay).TranslationY = -18
                DirectCast(secondChildOverlay, IAgStkGraphicsOverlay).ClipToParent = False
                secondChildOverlay.Texture = texture2D

                overlayManager.Add(DirectCast(overlay, IAgStkGraphicsScreenOverlay))

                Dim parentOverlayManager As IAgStkGraphicsScreenOverlayCollectionBase = TryCast(DirectCast(overlay, IAgStkGraphicsOverlay).Overlays, IAgStkGraphicsScreenOverlayCollectionBase)
                parentOverlayManager.Add(DirectCast(childOverlay, IAgStkGraphicsScreenOverlay))

                Dim childOverlayManager As IAgStkGraphicsScreenOverlayCollectionBase = TryCast(DirectCast(childOverlay, IAgStkGraphicsOverlay).Overlays, IAgStkGraphicsScreenOverlayCollectionBase)
                childOverlayManager.Add(DirectCast(secondChildOverlay, IAgStkGraphicsScreenOverlay))

                '#End Region

                m_Overlay = DirectCast(overlay, IAgStkGraphicsScreenOverlay)
                DirectCast(root, IAgAnimation).PlayForward()

            Catch e As Exception
                If e.Message.Contains("ProjectionRasterStreamPlugin") Then
                    MessageBox.Show("A COM exception has occurred." & vbLf & vbLf & _
                                    "It is possible that one of the following may be the issue:" & vbLf & vbLf & _
                                    "1. ProjectionRasterStreamPlugin.dll is not registered for COM interop." & vbLf & vbLf & _
                                    "2. That the plugin has not been added to the GfxPlugin category within a <install dir>\Plugins\*.xml file." & vbLf & vbLf & _
                                    "To resolve either of these issues:" & vbLf & vbLf & _
                                    "1. To register the plugin, open a Visual Studio " & _
                                    If(IntPtr.Size = 8, "x64 ", "") & "Command Prompt and execute the command:" & vbLf & vbLf & _
                                    vbTab & "regasm /codebase ""<install dir>\<CodeSamples>\Extend\Graphics\VB.Net\ProjectionRasterStreamPlugin\bin\<Config>\ProjectionRasterStreamPlugin.dll""" & vbLf & vbLf & _
                                    vbTab & "Note: if you do not have access to a Visual Studio Command Prompt regasm can be found here:" & vbLf & _
                                    vbTab & "C:\Windows\Microsoft.NET\Framework" & If(IntPtr.Size = 8, "64", "") & "\<.NET Version>\" & vbLf & vbLf & _
                                    "2. To add it to the GfxPlugins plugins registry category:" & vbLf & vbLf & _
                                    vbTab & "a. Copy the Graphics.xml from the <install dir>\CodeSamples\Extend\Graphics\Graphics.xml file to the <install dir>\Plugins directory." & vbLf & vbLf & _
                                    vbTab & "b. Then uncomment the plugin entry that contains a display name of ProjectionRasterStreamPlugin.VBNET." & vbLf & vbLf, "Plugin Not Registered", MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
                Else
                    MessageBox.Show("Could not create globe overlay.  Your video card may not support this feature.", _
                                    "Unsupported", MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
                End If
                Return
            End Try
        End Sub

		Public Overrides Sub View(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
			'
			' Overlays are always fixed to the screen regardless of view
			'
			scene.Render()
		End Sub

		Public Overrides Sub Remove(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
			If m_Overlay IsNot Nothing Then
				Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager
				Dim overlayManager As IAgStkGraphicsScreenOverlayCollectionBase = DirectCast(manager.ScreenOverlays.Overlays, IAgStkGraphicsScreenOverlayCollectionBase)

				DirectCast(root, IAgAnimation).Rewind()
				overlayManager.Remove(m_Overlay)
				scene.Render()

				m_Overlay = Nothing
			End If
		End Sub

		Private m_Overlay As IAgStkGraphicsScreenOverlay
	End Class
End Namespace

File: CustomApplications/CSharp/GraphicsHowTo/ScreenOverlays/OverlaysPanelCodeSnippet.cs
#region UsingDirectives

using System.Text;
using System.IO;
using System.Drawing;
using GraphicsHowTo.Imaging;
using AGI.STKGraphics;
using AGI.STKObjects;
using System;
using System.Windows.Forms;

#endregion

namespace GraphicsHowTo.ScreenOverlays
{
    public class OverlaysPanelCodeSnippet : CodeSnippet
    {
        public OverlaysPanelCodeSnippet()
            : base(@"ScreenOverlays\OverlaysPanelCodeSnippet.cs")
        {
        }

        public override void Execute(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            string imageFile = new AGI.DataPath(AGI.DataPathRoot.Relative, "Markers/originalLogo.png").FullPath;
            string rasterFile = new AGI.DataPath(AGI.DataPathRoot.Relative, "Textures/SpinSat_256.gif").FullPath;
            Execute(scene, root, imageFile, rasterFile);
        }

        [AGI.CodeSnippets.CodeSnippet(
            /* Name        */ "AddAPanelOverlay",
            /* Description */ "Add overlays to a panel overlay",
            /* Category    */ "Graphics | ScreenOverlays",
            /* References  */ "System",
            /* Namespaces  */ "System",
            /* EID         */ "AgSTKGraphicsLib~IAgStkGraphicsScreenOverlay"
            )]
        public void Execute([AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")] IAgStkGraphicsScene scene, [AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")] AgStkObjectRoot root, [AGI.CodeSnippets.CodeSnippet.Parameter("imageFile", "The image file to use for the child overlay")] string imageFile, [AGI.CodeSnippets.CodeSnippet.Parameter("rasterFile", "The raster file for the second child overlay")] string rasterFile)
        {
            try
            {
                #region CodeSnippet
                IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;
                IAgStkGraphicsScreenOverlayCollectionBase overlayManager = (IAgStkGraphicsScreenOverlayCollectionBase)manager.ScreenOverlays.Overlays;

                IAgStkGraphicsTextureScreenOverlay overlay =
                    manager.Initializers.TextureScreenOverlay.InitializeWithXYWidthHeight(/*$xLocation$The x location of the screen overlay$*/0, /*$yLocation$The y location of the screen overlay$*/0,
                    /*$overlayWidth$The width of the screen overlay$*/188, /*$overlayHeight$The height of the screen overlay$*/200);
                ((IAgStkGraphicsOverlay)overlay).Origin = /*$origin$The origin of the screen overlay$*/AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginTopLeft;
                ((IAgStkGraphicsOverlay)overlay).Color = /*$color$The System.Drawing.Color of the screen overlay$*/Color.LightSkyBlue;
                ((IAgStkGraphicsOverlay)overlay).Translucency = /*$translucency$The translucency of the screen overlay$*/0.7f;
                ((IAgStkGraphicsOverlay)overlay).BorderTranslucency = /*$borderTranslucency$The translucency of the border$*/0.3f;
                ((IAgStkGraphicsOverlay)overlay).BorderSize = /*$borderWidth$The width of the border$*/1;
                ((IAgStkGraphicsOverlay)overlay).BorderColor = /*$borderColor$The System.Drawing.Color of the border$*/Color.LightBlue;

                IAgStkGraphicsTextureScreenOverlay childOverlay =
                    manager.Initializers.TextureScreenOverlay.InitializeWithXYWidthHeight(
                    /*$childXLocation$The x location of the child overlay$*/0, /*$childYLocation$The y location of the child overlay$*/0, ((IAgStkGraphicsOverlay)overlay).Width, ((IAgStkGraphicsOverlay)overlay).Height);
                ((IAgStkGraphicsOverlay)childOverlay).Origin = /*$childOrigin$The origin of the child overlay$*/AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginCenter;

                childOverlay.Texture = manager.Textures.LoadFromStringUri(imageFile);

                //
                // Create the RasterStream from the plugin the same way as in ImageDynamicCodeSnippet.cs
                //
                IAgStkGraphicsProjectionRasterStreamPluginActivator activator =
                    manager.Initializers.ProjectionRasterStreamPluginActivator.Initialize();
                IAgStkGraphicsProjectionRasterStreamPluginProxy proxy =
                    activator.CreateFromDisplayName(/*$pluginDisplayName$Display Name of the ProjectionRasterStreamPlugin$*/"ProjectionRasterStreamPlugin.CSharp");
                Type plugin = proxy.RealPluginObject.GetType();
                plugin.GetProperty("RasterPath").SetValue(proxy.RealPluginObject, rasterFile, null);

                IAgStkGraphicsRasterStream rasterStream = proxy.RasterStream;
                rasterStream.UpdateDelta = /*$updateDelta$The interval at which the update method will be called$*/0.01667;
                IAgStkGraphicsRendererTexture2D texture2D = manager.Textures.FromRaster((IAgStkGraphicsRaster)rasterStream);

                IAgStkGraphicsTextureScreenOverlay secondChildOverlay =
                    manager.Initializers.TextureScreenOverlay.InitializeWithXYWidthHeight(/*$secondChildXLocation$The x location of the second child overlay$*/0, /*$secondChildYLocation$The y location of the screen overlay$*/0,
                    /*$secondChildWidth$The width of the second child overlay$*/128, /*$secondChildHeight$The height of the second child overlay$*/128);
                ((IAgStkGraphicsOverlay)secondChildOverlay).Origin = /*$secondChildOrigin$The origin of the second child overlay$*/AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginTopRight;
                ((IAgStkGraphicsOverlay)secondChildOverlay).TranslationX = /*$secondChildXTranslation$The X translation of the second child overlay$*/-36;
                ((IAgStkGraphicsOverlay)secondChildOverlay).TranslationY = /*$secondChildYTranslation$The Y translation of the second child overlay$*/-18;
                ((IAgStkGraphicsOverlay)secondChildOverlay).ClipToParent = /*$secondChildClipToParent$Whether or not the second child overlay should clip to its parent$*/false;
                secondChildOverlay.Texture = texture2D;

                overlayManager.Add((IAgStkGraphicsScreenOverlay)overlay);

                IAgStkGraphicsScreenOverlayCollectionBase parentOverlayManager = ((IAgStkGraphicsOverlay)overlay).Overlays as IAgStkGraphicsScreenOverlayCollectionBase;
                parentOverlayManager.Add((IAgStkGraphicsScreenOverlay)childOverlay);

                IAgStkGraphicsScreenOverlayCollectionBase childOverlayManager = ((IAgStkGraphicsOverlay)childOverlay).Overlays as IAgStkGraphicsScreenOverlayCollectionBase;
                childOverlayManager.Add((IAgStkGraphicsScreenOverlay)secondChildOverlay);
                #endregion

                m_Overlay = (IAgStkGraphicsScreenOverlay)overlay;
                ((IAgAnimation)root).PlayForward();
            }
            catch (Exception e)
            {
                if (e.Message.Contains("ProjectionRasterStreamPlugin"))
                {
                    StringBuilder sb = new StringBuilder();
                    sb.Append("A COM exception has occurred.\n\n");
                    sb.Append("It is possible that one of the following may be the issue:\n\n");
                    sb.Append("1. ProjectionRasterStreamPlugin.dll is not registered for COM interop.\n\n");
                    sb.Append("2. That the plugin has not been added to the GfxPlugin category within a <install dir>\\Plugins\\*.xml file.\n\n");
                    sb.Append("To resolve either of these issues:\n\n");
                    sb.Append("1. To register the plugin, open a Visual Studio ");
                    if (IntPtr.Size == 8)
                        sb.Append("x64 ");
                    sb.Append("Command Prompt and execute the command:\n\n");
                    sb.Append("\tregasm /codebase \"<install dir>\\<CodeSamples>\\Extend\\Graphics\\CSharp\\ProjectionRasterStreamPlugin\\bin\\<Config>\\ProjectionRasterStreamPlugin.dll\"\n\n");
                    sb.Append("\tNote: if you do not have access to a Visual Studio Command Prompt regasm can be found here:\n");
                    sb.Append("\tC:\\Windows\\Microsoft.NET\\Framework");
                    if (IntPtr.Size == 8)
                        sb.Append("64");
                    sb.Append("\\<.NET Version>\\\n\n");
                    sb.Append("2. To add it to the GfxPlugins plugins registry category:\n\n");
                    sb.Append("\ta. Copy the Graphics.xml from the <install dir>\\CodeSamples\\Extend\\Graphics\\Graphics.xml file to the <install dir>\\Plugins directory.\n\n");
                    sb.Append("\tb. Then uncomment the plugin entry that contains a display name of ProjectionRasterStreamPlugin.CSharp.\n\n");

                    MessageBox.Show(sb.ToString(), "Plugin Not Registered", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                }
                else
                {
                    MessageBox.Show("Could not create overlay.  Your video card may not support this feature.",
                        "Unsupported", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                }
                return;
            }
        }

        public override void View(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            //
            // Overlays are always fixed to the screen regardless of view
            //
            scene.Render();
        }

        public override void Remove(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            if (m_Overlay != null)
            {
                IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;
                IAgStkGraphicsScreenOverlayCollectionBase overlayManager = (IAgStkGraphicsScreenOverlayCollectionBase)manager.ScreenOverlays.Overlays;

                ((IAgAnimation)root).Rewind();
                overlayManager.Remove(m_Overlay);
                scene.Render();

                m_Overlay = null;
            }
        }

        private IAgStkGraphicsScreenOverlay m_Overlay;
    }
}